src/main/java/frc/robot/subsystems/conveyor/Conveyor.java

```diff
@@ -0,0 +1,253 @@
+        return INSTANCE;
+    }
+
+    /**
+     * @return the color sensor value as a DriverStation.Alliance enum.
+     */
+    public DriverStation.Alliance getColor() {
+        Color color = colorSensor.getColor();
+        ColorMatchResult result = match.matchClosestColor(color);
+        Color resultColor = result.color;
+        SmartDashboard.putBoolean("match", resultColor == Color.kRed);
+
+        if (resultColor == Constants.Conveyor.RED) {
+            return DriverStation.Alliance.Red;
+        } else if (resultColor == Constants.Conveyor.BLUE) {
+            return DriverStation.Alliance.Blue;
+        } else {
+            return DriverStation.Alliance.Invalid;
+        }
+    }
+
+    /**
+     * @return the amount of balls inside the cargo.
+     */
+    public int getCargoCount() {
+        return cargoPositions.size();
+    }
+
+    /**
+     * Set motor power.
+     *
+     * @param power the power to the motor [-1, 1].
+     */
+    public void setPower(double power) {
+        motor.set(power);
+    }
+
+    /**
+     * open the "flap" - solenoid
+     */
+    public void openFlap() {
+        flap.set(false);
+    }
+
+    /**
+     * closes the "flap" - solenoid
+     */
+    public void closeFlap() {
+        flap.set(true);
+    }
+
+    /**
+     * toggles the "flap" - solenoid
+     */
+    public void toggleFlap() {
+        flap.toggle();
+    }
+
+    private void updateActualBallPositions() throws InterruptedException {
+        var colorIntake = getColor();
+        boolean isPostFlapBeamActive = postFlapBeam.get();
+        SmartDashboard.putString("alliance", colorIntake.name());
+        double power = motor.getMotorOutputPercent();
+        /*
+        Check whether the post flap beam input changes and the conveyor is moving balls in
+            true => Check whether the last value is invalid
+                true => Take out the head of the queue
+                false => Take out the value in the queue
+
+        Check whether color sensor input changes and the current input isn't invalid
+            true => Check whether the conveyor is moving balls out
+                true => Check whether there are any remaining spaces in the queue
+                    true => Take out the head of the queue
+                    Add a ball to the end of the queue
+                false => Check whether the conveyor is moving balls in
+                    Remove the head of the queue
+                    Add another ball at the tail of the queue
+         */
+        if (isPostFlapBeamActive && !wasPostFlapBeamActive && power > 0) {
+            cargoPositions.removeFirst();
+            cargoPositions.add(DriverStation.Alliance.Invalid.name());
+        }
+        if (colorIntake != lastSeenColor && !colorIntake.equals(DriverStation.Alliance.Invalid) && power > 0) {
+            cargoPositions.removeFirstOccurrence(DriverStation.Alliance.Invalid.name());
+            cargoPositions.add(colorIntake.name());

```

Comment:
What happens if the queue looks like this. [alliance, opponent]

Reply:
Well that's the problem of the driver.